include/tvm/compilation.h

```diff
@@ -0,0 +1,211 @@
+    std::string str() const {
+        std::ostringstream result;
+        result << targetName;
+        for (const auto &x : options) {
+            result << " " << x;
+        }
+        return result.str();
+    }
+
+    /*! \return True iff this target has the given key */
+    bool hasKey(std::string key) const {
+        return keys.count(key) > 0;
+    }
+};
+
+/*! \return A target for LLVM */
+Target target_llvm() {
+    std::unordered_set<std::string> keys({ "llvm", "cpu" });
+    std::unordered_set<std::string> options;
+    return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+/*! \return A target for CUDA */
+Target target_cuda() {
+    std::unordered_set<std::string> keys({ "cuda", "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("cuda", kDLGPU, 512, 32, keys, options);
+}
+
+/*! \return A target for ROCm */
+Target target_rocm() {
+    std::unordered_set<std::string> keys({ "rocm", "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("rocm", kDLROCM, 256, 1, keys, options);
+}
+
+/*! \return A target for Metal */
+Target target_metal() {
+    std::unordered_set<std::string> keys({ "gpu" });
+    std::unordered_set<std::string> options;
+    return Target("metal", kDLMetal, 256, 1, keys, options);
+}
+
+/*! \return A target for rasp */
+Target target_rasp() {
+    std::unordered_set<std::string> keys({ "llvm", "cpu" });
+    std::unordered_set<std::string> options({
+        "-device=rasp",
+        "-mtriple=armv7l-none-linux-gnueabihf",
+        "-mcpu=cortex-a53",
+        "-mattr=+neon"
+    });
+    return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+/*! \return A target for stackvm */
+Target target_stackvm() {
+    std::unordered_set<std::string> keys({ "stackvm", "cpu" });
+    std::unordered_set<std::string> options;
+    return Target("stackvm", kDLCPU, 512, 1, keys, options);
+}
+
+
+bool LLVMEnabled() {
+    const runtime::PackedFunc* pf = runtime::Registry::Get("codegen.llvm_target_enabled");
+    if (pf == nullptr) return false;
+    return (*pf)("llvm");
+}
+
+/*! \return The default host target for a given device target */
+Target default_target_host(Target target) {
+    if (target.deviceType == kDLCPU) {
+        return target;
+    } else {
+        if (LLVMEnabled()) {
+            return target_llvm();
+        } else {
+            return target_stackvm();
+        }
+    }
+}
+
+
+/*!
+* \brief Container for build configuration options
+*/
+struct BuildConfig {
+    int data_alignment = -1;
+    int offset_factor = 0;
+
+    int double_buffer_split_loop = 1;
+    int auto_unroll_max_step = 0;
+    int auto_unroll_max_depth = 8;
+    int auto_unroll_max_extent = 0;
+    bool unroll_explicit = true;
+
+    bool restricted_func = true;
+
+    bool detect_global_barrier = false;
+
+    BuildConfig() {
+    }
+};
+
+
+/*! \brief Convenience function for getting attributes */
+TVMValue GetAttr(NodeRef node, std::string attrName) {
```

Comment:
This function do not belong to here. Instead of using GetAttr, use 

```node.as<Expr>->type()``` for the type